### ISSUE TITLE
Simultaneous SD2 and EventAI scripts

### ISSUE THREAD

**Issue by ccjara** [scriptdev2/scriptdev2#189]
I am working on a dungeon event where only specific mobs are coupled to it. There are some event triggers which (to my understanding) cannot be handled by the DB because they require some core logic, such as being close to the player, initialisation on area triggering and also more exotic actions, such as rotating, changing faction and stand state.

I registered both EventAI and mob_xyz in the "AiName" and "ScriptName" column, however it seems that the SD2 script is overwriting the EventAI. As soon as I register the SD2 script, the casters are just melee hitting (since I am overwriting the ScriptedAI::Update function), while the SD2 script is running without problems. On the other hand, when I don't register the SD2 script, the EventAI works fine.

Is it possible to have both AI systems running at the same time?

In short I want to extend the existing EventAI script with an SD2 one on triggers that can only be recognised on the SD2 side. The resulting event action can only be executed by SD2 as well.

**Comment by xfurry** [comment]
You can't have the AI component in both places. 
The only option is to have a  spell or gossip script in SD2 and AI script in ACID.

**Comment by ccjara** [comment]
**Edit:** I just stumpled upon EVENT_T_OOC_LOS. This might actually be the way to do it. Thanks for hinting me that way.

**Comment by ccjara** [comment]
Okay, triggering it actually works fine, but I cannot figure out how to send a signal to SD2, including the creature reference. There is no spell the creature uses that I could hook on using scripted_spell. Any idea what I can do?

**Comment by xfurry** [comment]
Send AI event function in eventAI

**Comment by ccjara** [comment]
Doesn't this send an event to ACID from SD2? I need it the other way around because the combat AI should remain in ACID.

A nice way would be

Event: EVENT_T_OOC_LOS, Action: ACTION_T_SEND_EVENT

which then I could catch in SD2 by ReceiveAiEvent()

**Comment by xfurry** [comment]
https://github.com/cmangos/mangos-wotlk/blob/master/doc/EventAI.txt#L153
https://github.com/cmangos/mangos-wotlk/blob/master/doc/EventAI.txt#L95

This function will throw an AI event from your creature to any other creature around, regardless if the other creature is scripted in SD2 or eventAI.

As an SD2 example you can look here: https://github.com/cmangos/mangos-wotlk/commit/1cfffe8045ed5598ba0f2e767d81d690336baac3#diff-724bb03c3ae5df3987bef4d64d9a4915R471

**Comment by ccjara** [comment]
Does it also throw the event to itself? Because I need to handle the event on the creature itself that threw the event.

1.) Player goes near creature
2.) This triggers ACID event EVENT_T_OOC_LOS
3.) Action is sent to SD2
4.) SD2 receives the thrown event via ReceiveAiEvent()
5.) SD2 performs some actions on the creature that threw the event which ACID is not able to do

**Comment by xfurry** [comment]
> Does it also throw the event to itself?

In eventAI, no. In SD2, yes, but it doesn't matter, as in SD2 you can do anything without such a function.

**Comment by ccjara** [comment]
But then again, in order to catch the event in SD2, I would need to register the mob script so I can make use of the ReceiveAiEvent function which would overwrite the ACID script from the beginning. Or did I misunderstood something there?

Is there a way to relay it to the ScriptedInstance so I do not need to register an SD2 script for the trash mobs?

I can only think of ScriptedInstance::SetData, but this is rather used for GOs and encounter states. Seems hacky to implement it that way.

**Comment by ghost** [comment]
``` QUOTE
Does it also throw the event to itself?
```

SD2 - yes.
EventAI - no.
Db scripts - yes.

**Comment by ccjara** [comment]
I thought about the issue a lot and unfortunately I cannot find a way to fix it. I will explain the event more detailed to clarify the problem I am facing.

I am currently working on the Corrupted Ashbringer event. When you enter the Cathedral instance an area trigger fires. The script then checks whether there is a wielder of the Corrupted Ashbringer and in case, all players of the party receive a _temporary_ friendly reputation with the Scarlet Crusade, on top of an instance wide yelling of Morgraine. When the Ashbringer wielder approaches any creature inside the instance, they face them, wait for 2 seconds, then kneel and say a random prepared text.

There are also 2 special boss interactions. An event between Morgraine and his father's ghost and a gossip script with Fairbanks. I already completed these (and also the trashmob scripts, areatriggers etc.). They were easy to script since I was simply extending  / creating the SD2 code.

But there is this problem with the regular trashmobs I can't get to fix. The NPCs already have an ACID script for combat interaction so I cannot give them an SD2 script, since it overwrites ACID on self registering, but at the same time ACID is unaware of the event itself and the functionality is too limited.

It would be nice if it was possible to simply inject an SD2 script for all the trashmobs in case the Ashbringer event has started.

Maybe you are more experience with an issue like this and have an idea how to fix it?

**Comment by Phatcat** [comment]
.... You could move the contents of the acid files for the trash mobs into their own sd2 scripts and expand on them there.... Dunno if this would be the prefered way to go, though, but it could be a way to go.

**Comment by Phatcat** [comment]
@harlynx If you are looking at the corrupted ashbringer event, you may also want to look into what is causing unrelated mobs to go hostile with you. You can reproduce the issue by equipping the corrupted ashbringer, then tele to westfall and seek out some defias mobs. Or, just seek out any critter and stand next to it for a short while.

**Comment by xfurry** [comment]
@harlynx does something like this help: https://github.com/cmangos/mangos-wotlk/blob/master/src/scriptdev2/scripts/eastern_kingdoms/stormwind_city.cpp#L616 ?

**Comment by ccjara** [comment]
@Phatcat Yes that would have been my last way of doing it. I was hoping that it would be possible without altering the DB in such a way.

Yes I noticed that the first time when the bunnys would go hostile on me when wearing the weapon. The Spell 28414 is applied when you equip the weapon. One of its effects is to periodically proc a dummy spell 28441 every 5 seconds. I don't know for sure what it is for, but I am certain is not for the event in the SM:C dungeon as it does not affect friendly creatures. I will try to find a fix while I am at it but I think it is core related.

@xfurry Yes I have a similar implementation (mob -> player). Like your example, it requires a mob script to register which overwrites the ACID script.

If there aren no other ways, I will implement it as Phatcat suggested. For me the best solution would have been to extend the core to allow additional scripts to be executed for a creature at runtime.

If another or better solution comes to your mind, feel free to tell me.

**Comment by Phatcat** [comment]
@harlynx The reason I mentioned defias mobs is because they will all start yelling that you've betrayed them and stuff like that (executing speech texts that are supposed to be exclusive to Argent Dawn). Does all humanoid mobs do this? - Also if it is because of the triggering of a dummy effect which does it, then it is surely a core issue as dummies are handled in the core.

*Edit: My guess for what the dummy effect is for is for rolling on a chance to get one of those 'periodic, yet random whispers' that would go along with having the weapon equipped. But that's just a guess.

**Comment by ccjara** [comment]
Ahh yes I totally forgot about the Argent Dawn part. :s

Usually other creatures should not be affected by it. I just tested it again. The dummy spell basically aggroes everything that is not friendly, which is a bug for sure.

Edit: I could not reproduce that the defias mobs would yell Argent Dawn texts.

**Comment by Phatcat** [comment]
@harlynx Here is a great video on the whole event actually: https://www.youtube.com/watch?v=Fea71ZhDzZ4 (Just ignore the shit going beyond wrath)

**Comment by ccjara** [comment]
@Phatcat Yes, I used that among dozens of other people's literal "walkthroughs" through the event. Thanks. :+1:

**Comment by xfurry** [comment]
@harlynx actually the logic of what you are trying to do is pretty simple.
You have spell http://www.wowhead.com/spell=28441/ab-effect-000 triggered by http://www.wowhead.com/spell=28282/ashbringer
You can have script targets on 28441 and have some DBscript or SD2 script that will achieve the desired effect.

You can see an example here, https://github.com/scriptdev2/scriptdev2/commit/42c495f0c31db520d04818b39503117cfb80fe38

This piece of code doesn't involve adding an AI component in SD2. You just add the dummy spell part in SD2 and keep the AI in ACID.

**Comment by ccjara** [comment]
I already tried to use that spell, but there were two issues: first it only triggers every 5 seconds. On all the videos I have seen the mobs turn around immediately when the player is closer than ~10 yards. However that would be a tradeoff I would gladly take since porting 9 mob scripts over to SD2 just for the sake of the event is a nightmare.

Second, I don't know whether the spell effect will actually be triggered since friendly mobs are not affected by it, only neutral / hostile.

**Comment by Phatcat** [comment]
@harlynx No, that is because you need to not have aggroed them when it triggers. Easiest way to reproduce is stand outside of a hut or building, then just wait a bit, they will start yelling inside and come running out... let me just capture a screendump for you.

**Comment by Phatcat** [comment]
![defias_brotherhood](https://cloud.githubusercontent.com/assets/1907820/15615298/c3d656a2-243d-11e6-9e14-782ea689fa4a.jpg)

**Comment by ccjara** [comment]
What I meant is I was not able to reproduce that the Defias would yell Argent Dawn texts. Everything non-friendly aggroes me instantly - critters, NPCs, anything, as you said. The periodic spell is considered a harmful spell which aggroes them.

At the same time I wonder whether the AI recognizes the "spell hit" if the creature is friendly. It surely did not when I tested it on the EventAI.

**Comment by Phatcat** [comment]
I know that, which is why I provided you with the method that produced this issue for me, as well as a screenshot of it. It even raises another question which is; should these mobs be affected by the spell at all if they are not in a clear path to the caster? Because triggering mobs through buildings seem wrong to me, even for something so exceptional as this.

**Comment by xfurry** [comment]
> Second, I don't know whether the spell effect will actually be triggered since friendly mobs are not affected by it, only neutral / hostile.

@harlynx Spell `ID - 28441 AB Effect 000` has

```
Effect 0: Id 3 (SPELL_EFFECT_DUMMY)
BasePoints = 1
Targets (22, 7) (TARGET_CASTER_COORDINATES, TARGET_AREAEFFECT_INSTANT)
Radius (Id 18) 15.00
```

which means that you can add the following

``` sql
DELETE FROM spell_script_target WHERE entry IN (28441);
INSERT INTO spell_script_target (entry,type,targetEntry,inverseEffectMask) VALUES
(28441,1,x,0),
(28441,1,y,0),
-- and so on...
```

Where x, y are the creature entries that you want to be hit by the spell, within the spell radius.
Hence you can use friendly and hostile units, as per your needs.

Also, since this is a dummy effect, you can either write some SD2 code as per my example above, or you can use dbscripts_on_spell, link here: https://github.com/cmangos/issues/wiki/DBScripts to script your event.
As I can see here https://www.youtube.com/watch?v=m3E6W6-rRGA, you will probably need SCRIPT_COMMAND_SET_FACING and SCRIPT_COMMAND_EMOTE on the target, and it should be perfectly fine.

Also, as per the video, the timer of 5 sec looks perfectly fine, as they don't face the player instantly when the npc is in the range of 18.

Let me know if this works fine for you, and also, please provide your patch to the DB devs so we can also integrate this event into the full cmangos release.

**Comment by ccjara** [comment]
dbscripts_on_spell is actually great, however there are some issues:
1. There are only 4 fields to store values for the SCRIPT_COMMAND_TALK command, but the creatures can say up to 7 texts.
2. The event will repeat and there does not seem to be a way to stop that. I implemented a SCRIPT_COMMAND_TERMINATE_CONDITION check (which is needed) and also placed the condition "if not already kneeling" there, but some creatures inside the chapel already kneel. Is there a way to store a flag for the creature when the event successfully executed once?
3. Stopping movement of randomly walking creatures is not really working smoothly. They sometimes don't turn around or stop their movement.

Beforehand I tried to register a Spell Effect Dummy script in the instance script, which did not work. When I registered it in the mob script, it worked but as mentioned before, I can't use it there because the ACID script would be overwritten.

**Comment by xfurry** [comment]
> Beforehand I tried to register a Spell Effect Dummy script in the instance script, which did not work. When I registered it in the mob script, it worked but as mentioned before, I can't use it there because the ACID script would be overwritten.

It has to work, if you do it properly.
Take the example of creature http://www.wowhead.com/npc=25316/beryl-sorcerer
This creature has the AI scripted in ACID and the dummy spell in SD.

https://github.com/cmangos/mangos-wotlk/blob/master/src/scriptdev2/scripts/northrend/borean_tundra.cpp#L1181
You only need to make sure, **NOT** to include the `pNewScript->GetAI = &GetAI_npc_your_npc` statement in our code and have 

``` sql
 UPDATE `creature_template` SET `AIName` = 'EventAI' WHERE `entry` = xyz;
```

added to your creature along with the ACID script.

**Comment by ccjara** [comment]
Nice, that works. Thanks a lot. Will finish it and then submit it.
